src/main/java/org/openpnp/machine/reference/camera/Webcams.java

```diff
@@ -0,0 +1,251 @@
+
+
+
+import org.openpnp.CameraListener;
+import org.openpnp.gui.support.PropertySheetWizardAdapter;
+import org.openpnp.gui.support.Wizard;
+import org.openpnp.gui.wizards.CameraConfigurationWizard;
+import org.openpnp.machine.reference.ReferenceCamera;
+import org.openpnp.machine.reference.camera.wizards.WebcamConfigurationWizard;
+import org.openpnp.spi.PropertySheetHolder;
+import org.openpnp.util.OpenCvUtils;
+import org.simpleframework.xml.Attribute;
+
+import java.awt.Dimension;
+import java.io.File;
+import java.io.IOException;
+
+import javax.imageio.ImageIO;
+
+import com.github.sarxos.webcam.*;
+import com.github.sarxos.webcam.util.jh.JHGrayFilter;
+
+
+
+/**
+ * A Camera implementation based on the OpenCV FrameGrabbers.
+ */
+public class Webcams extends ReferenceCamera implements Runnable , WebcamImageTransformer {
+    
+	@Attribute(required=false)
+	protected String deviceId = "###DEVICE###";
+
+	@Attribute(required=false)
+	private int preferredWidth=0;
+	@Attribute(required=false)
+	private int preferredHeight=0;
+	
+	protected Webcam webcam;
+	private Thread thread;
+	private boolean test;
+	private BufferedImage image;
+
+	private static final JHGrayFilter GRAY = new JHGrayFilter();
+
+
+	@Override
+	public BufferedImage transform(BufferedImage image) {
+		return GRAY.filter(image, null);
+	}
+
+	Webcams() {
+	}
+	
+	@Override
+	public synchronized BufferedImage capture() {
+	    if (thread == null) {
+	        setDeviceId(deviceId);
+	    }
+		if(thread==null) { return null; }
+		try {
+		    BufferedImage img = webcam.getImage();
+		    return transformImage(img);
+		}
+		catch (Exception e) {
+			return null;
+		}
+	}
+	
+	@Override
+    public synchronized void startContinuousCapture(CameraListener listener, int maximumFps) {
+	    if (thread == null) {
+	        setDeviceId(deviceId);
+	    }
+        super.startContinuousCapture(listener, maximumFps);
+    }
+
+    private BufferedImage lastImage=null;
+    private BufferedImage redImage = new BufferedImage(1,1,BufferedImage.TYPE_INT_ARGB);
+					
+
+    public void run() {
+		while (!Thread.interrupted()) {
+			try {
+				BufferedImage image = capture();
+				if (image == null) {
+					image = redImage;
+				}
+					broadcastCapture(image);
+			}
+			catch (Exception e) {
+				e.printStackTrace();
+			}
+			try {
+				Thread.sleep(1000 / 30);
+			}
+			catch (InterruptedException e) {
+				break;
+			}
+		}
+	}
+	
+	public String getDeviceId() {
+		return deviceId;
+	}
+
+	public synchronized void setDeviceId(String deviceId) {
+		this.deviceId = deviceId;
+		if (thread != null) {
+			thread.interrupt();
+			try {
+				thread.join();
+			}
+			catch (Exception e) {
+				e.printStackTrace();
+			}
+			thread = null;
+			webcam.close();
+		}
+		try {
+		webcam=null;
+                for (Webcam cam: Webcam.getWebcams()) {
+                        if(cam.getName().equals(deviceId)) {
+				webcam=cam;
+			}
+		}
+		if(webcam==null) {	// Workaround ???
```

Comment:
Because i have used it during test, and at the end before submitting
changed it to gray,
same as on LtiCivilCamera, where the Wizard was copied. Should i
remove it completly,
or rename it to forceGray ?

For the comment:
       if(webcam==null) {      // Workaround ???
                        //webcam =  Webcam.getWebcamByName(deviceId);
                        return;
                }

It was a hint.
It's if the webcam is not found, should the driver open the first cam ,
   webcam = Webcam.getDefault()  ;

or should it as it works now, don't open the camera.




2016-02-29 17:18 GMT, Jason von Nieda <notifications@github.com>:
>> +				thread.join();
>> +			}
>> +			catch (Exception e) {
>> +				e.printStackTrace();
>> +			}
>> +			thread = null;
>> +			webcam.close();
>> +		}
>> +		try {
>> +		webcam=null;
>> +                for (Webcam cam: Webcam.getWebcams()) {
>> +                        if(cam.getName().equals(deviceId)) {
>> +				webcam=cam;
>> +			}
>> +		}
>> +		if(webcam==null) {	// Workaround ???
>
> This comment is confusing. Is it a TODO? Can you explain or remove?
>
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/openpnp/openpnp/pull/238/files#r54443332